x-pack/plugins/files/server/file/file.ts

```diff
@@ -0,0 +1,161 @@
+  File as IFile,
+  FileSavedObject,
+  FileSavedObjectAttributes,
+  FileStatus,
+  UpdatableFileAttributes,
+} from '../../common';
+import { BlobStorageService } from '../blob_storage_service';
+import {
+  fileAttributesReducer,
+  Action,
+  createDefaultFileAttributes,
+} from './file_attributes_reducer';
+import { createAuditEvent } from '../audit_events';
+import { InternalFileService } from '../file_service/internal_file_service';
+
+/**
+ * Public class that provides all data and functionality consumers will need at the
+ * individual file level
+ */
+export class File<M = unknown> implements IFile {
+  private readonly logAuditEvent: InternalFileService['createAuditLog'];
+
+  constructor(
+    private fileSO: FileSavedObject,
+    private readonly internalFileService: InternalFileService,
+    private readonly blobStorageService: BlobStorageService,
+    private readonly logger: Logger
+  ) {
+    this.logAuditEvent = this.internalFileService.createAuditLog.bind(this.internalFileService);
+  }
+
+  private async updateFileState(action: Action) {
+    this.fileSO = await this.internalFileService.updateSO(
+      this.id,
+      fileAttributesReducer(this.attributes, action)
+    );
+  }
+
+  private canUpload(): boolean {
+    return (
+      this.fileSO.attributes.status === 'AWAITING_UPLOAD' ||
+      this.fileSO.attributes.status === 'UPLOAD_ERROR'
+    );
+  }
+
+  public async update(attrs: UpdatableFileAttributes): Promise<IFile> {
+    await this.updateFileState({
+      action: 'updateFile',
+      payload: attrs,
+    });
+    return this;
+  }
+
+  public async uploadContent(content: Readable): Promise<void> {
+    if (!this.canUpload()) {
+      this.logger.error('File content already uploaded.');
+      throw new Error('File content already uploaded');
+    }
+    this.logger.debug('Uploading file contents...');
+    await this.updateFileState({
+      action: 'uploading',
+    });
+
+    try {
+      const { id: contentRef, size } = await this.blobStorageService.upload(content);
+      await this.updateFileState({
+        action: 'uploaded',
+        payload: { content_ref: contentRef, size },
+      });
+    } catch (e) {
+      await this.updateFileState({ action: 'uploadError' });
+      throw e;
+    }
+  }
+
+  public downloadContent(): Promise<Readable> {
+    const { content_ref: id, size } = this.attributes;
+    if (!id) {
+      throw new Error('No content to download');
+    }
+    return this.blobStorageService.download(id, size);
+  }
+
+  public async delete(): Promise<void> {
+    const { attributes, id } = this.fileSO;
+    if (attributes.content_ref) {
+      await this.blobStorageService.delete(attributes.content_ref);
+    }
+    await this.internalFileService.deleteSO(id);
+    this.logAuditEvent(
+      createAuditEvent({
+        action: 'delete',
+        outcome: 'success',
+        message: `Deleted file "${this.name}" of kind "${this.fileKind}" with id "${this.id}"`,
+      })
+    );
+  }
```

Comment:
I'm thinking, to keep consistency. We might want to introduce `DELETED` state in our `status` field. And the deletion flow would go as follows.

If we don't support file sharing across spaces:

1. Set file status in the SO to `DELETED`.
2. Delete blob contents.
3. Delete file SO.

If we support file sharing across spaces:

1. Set file status in the SO to `DELETED`.
2. Background task would eventually pick up this SO.
   1. Check if there are no other non-deleted SO linking to the blob contents.
      1. Delete blob in blob storage.
      2. Delete SO.